src/services/meta_transaction_service.ts

```diff
@@ -44,223 +50,140 @@ import { ethGasStationUtils } from '../utils/gas_station_utils';
-                new BigNumber(CHAIN_ID),
-                this._contractWrappers.contractAddresses.exchange,
-            )
-            .callAsync();
-
-        const makerAssetAmount = swapQuote.bestCaseQuoteInfo.makerAssetAmount;
-        const totalTakerAssetAmount = swapQuote.bestCaseQuoteInfo.totalTakerAssetAmount;
-        const allowanceTarget = this._contractAddresses.erc20Proxy;
-        const apiMetaTransactionQuote: GetMetaTransactionQuoteResponse = {
-            price,
+    ): Promise<GetMetaTransactionQuoteResponseV0 | GetMetaTransactionQuoteResponseV1> {
+        const { swapVersion } = params;
+        const quote = await this._calculateMetaTransactionQuoteAsync(params);
+        const commonQuoteFields = {
+            price: quote.price,
             sellTokenAddress: params.sellTokenAddress,
             buyTokenAddress: params.buyTokenAddress,
-            zeroExTransactionHash,
-            zeroExTransaction,
-            buyAmount: makerAssetAmount,
-            sellAmount: totalTakerAssetAmount,
-            orders: serviceUtils.cleanSignedOrderFields(orders),
-            sources: serviceUtils.convertSourceBreakdownToArray(sourceBreakdown),
-            gasPrice,
-            estimatedGas,
-            gas: estimatedGas,
-            protocolFee,
-            minimumProtocolFee,
+            buyAmount: quote.buyAmount,
+            sellAmount: quote.sellAmount,
+            orders: quote.orders,
+            sources: quote.sources,
+            gasPrice: quote.gasPrice,
+            estimatedGas: quote.estimatedGas,
+            gas: quote.estimatedGas,
+            protocolFee: quote.protocolFee,
+            minimumProtocolFee: quote.minimumProtocolFee,
             estimatedGasTokenRefund: ZERO,
-            value: protocolFee,
-            allowanceTarget,
+            value: quote.protocolFee,
+            allowanceTarget: quote.allowanceTarget,
+        };
+
+        if (swapVersion === SwapVersion.V0) {
+            // Go through the Exchange.
+            const emtx = this._generateExchangeMetaTransaction(
+                quote.orders,
+                quote.sellAmount,
+                quote.buyAmount,
+                quote.orders.map(order => order.signature),
+                quote.takerAddress,
+                normalizeGasPrice(quote.gasPrice),
+            );
+            return {
+                ...commonQuoteFields,
+                zeroExTransaction: emtx,
+                zeroExTransactionHash: getExchangeMetaTransactionHash(emtx),
+            };
+        }
+        // Go through the Exchange Proxy.
+        const epmtx = this._generateExchangeProxyMetaTransaction(
+            quote.callData,
+            quote.takerAddress,
+            normalizeGasPrice(quote.gasPrice),
+            calculateProtocolFeeRequiredForOrders(swapVersion, normalizeGasPrice(quote.gasPrice), quote.orders),
+        );
+        return {
+            ...commonQuoteFields,
+            mtx: epmtx,
```

Comment:
are we using `mtx` to determine if its EP or not? Don't love it if `mtx` leaks to the user

Reply:
Then endpoint (`v0` or `v1`) determines the type of mtx. What's being leaked? `v0` returned the meta-transaction object also?